### PR TITLE
cluster-handling kube caches

### PR DESCRIPTION
**Pull request by joekelley**
Wrappers for kube caches to enable sharing caches across clusters, per-client.
BOT NOTES: 
resolves https://github.com/solo-io/solo-kit/issues/295

### CODE REVIEW

**Comment by ashleywang1** [comment]
Wait, before you merge this, would it be possible to change:
`func NewKubeCoreCache(ctx context.Context, client kubernetes.Interface) (*kubeCoreCaches, error) {`
to 
`func NewKubeCoreCache(ctx context.Context, client kubernetes.Interface) (KubeCoreCache, error) {` ?

That makes it easier for wire.go to deal with it (fixes the wire injection issue I had in sm-marketplace)

File: pkg/multicluster/clustercache/cache_manager.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+//go:generate mockgen -destination=./mocks/cache_manager.go -source cache_manager.go -package mocks
+
+type cacheWrapper struct {
+	cancel    context.CancelFunc
+	coreCache PerClusterCache
+}
+
+type PerClusterCache interface {
+	IsPerCluster()
```
Comment (ilackarms) [inline pkg/multicluster/clustercache/cache_manager.go]: can you add some comments here to describe the intent / effect of this interface?

File: pkg/multicluster/clustercache/cache_manager.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+type FromConfig func(ctx context.Context, cluster string, restConfig *rest.Config) PerClusterCache
```
Comment (ilackarms) [inline pkg/multicluster/clustercache/cache_manager.go]: nit: rename `FromConfig` to something more descriptive

File: pkg/multicluster/clustercache/cache_manager.go (hunk 1 above)
Comment (ilackarms) [reply pkg/multicluster/clustercache/cache_manager.go]: nit: consider renaming `PerClusterCache` to `ClusterCache` and `IsPerCluster()` to `IsClusterCache()`

**Pull request merged**